imap/mboxlist.c

```diff
@@ -5423,6 +5423,21 @@ static int _foreach_cb(void *rock,
     mbentry->name = xstrndup(key, keylen);
     mbentry->mbtype |= MBTYPE_LEGACY_DIRS;
 
+    if (!mbentry->uniqueid) {
+        /* Fetch uniqueid from cyrus.header */
+        struct mailbox *mailbox = NULL;
+        int r = mailbox_open_from_mbe(mbentry, &mailbox);
+        if (r) {
+            mboxlist_entry_free(&mbentry);
+            return r;
```

Comment:
I think this needs to log the error and return 0 rather than r.  Otherwise the cyrusdb_foreach will bail out at the first problem, and whatever mbentries we didn't get to will be missing from the upgraded db later.  Same applies at the end of this function (which doesn't change in this diff).  The overall return value is eventually ignored by process_mboxlist() anyway, so it will look like it succeeded, even if it craps out early and does very little.

Reply:
So, if we fail to open the mailbox, do we assume that is doesn't exists and ignore the mbentry?

If we do bail, then the original mailboxes.db remains as-is.  We could/should report an error.

Reply:
It's a tricky one -- on one hand, maybe we should crap out completely (i.e. fail all the way up until we exit non-zero from ctl_cyrusdb, so that master doesn't even start services if any part of the upgrade didn't succeed).  The admin will be forced to report/fix whatever is preventing that mailbox from opening, and try again.

On the other hand, if a mailbox fails to open, is that even a class of error that the sysadmin can repair?  Especially in a state where Cyrus can't start up and the cmdline tools can't function because the upgrade hasn't succeeded yet!  I guess it would depend on why it failed to open, which we don't know yet -- it's a hypothetical edge-case question at the moment.

I think for now, the best behaviour is "do everything we possibly can, but log and skip individual mailboxes that don't work for some reason", and that means returning zero even when they fail, so we at least continue with the rest.  I imagine this is probably what you've just done, I haven't looked at your latest updates yet.

I guess the next 3.6 beta will give us some data about what sort of cases, if any, it still trips on, and then having that data we can work out how to deal with it.

Reply:
What do we do if  we fail to parse the mbentry?  Try to piece one together from the mailbox header, or log and skip?

Reply:
Log and skip too, I think -- at least for now.  Maybe if we see a bunch of failure patterns that are obviously recoverable we can think about whether it's worth it, otherwise let it fall through to reconstruct to handle after the fact.

That being said, I'm not sure how much reconstruct can currently handle -- first it needs the upgrade to succeed and produce a good database, so in my previous work it didn't have anything to work with.  But once this PR is merged, with the upgrade improvements and the new `mailbox_open_mbe()` API, I think I should be able to forward-port my reconstruct and ctl_cyrusdb -r work from 3.4 and have them actually function -- and if that turns out to be the case, then the upgrade skipping some individual entries it couldn't cope with won't be such a big deal, cause the admin can just reconstruct to get them back.

Reply:
Logging and skip is now implemented